global_helpers/panther_greynoise_helpers.py

```diff
@@ -63,114 +72,133 @@ def context(self, match_field) -> dict:
-
-    def cve_string(self, match_field, limit: int = 10) -> str:
-        cve_raw = deep_get(self.noise, match_field, "cve")
+    def cve_string(self, match_field: str, limit: int = 10) -> str:
+        cve_raw = self._lookup(match_field, "cve")
         if isinstance(cve_raw, list):
             return " ".join(cve_raw[:limit])
         return cve_raw
 
-    def cve_list(self, match_field) -> list:
-        cve_raw = deep_get(self.noise.get, match_field, "cve")
+    def cve_list(self, match_field: str) -> list:
+        cve_raw = self._lookup(match_field, "cve")
         if isinstance(cve_raw, str):
             return [cve_raw]
         return cve_raw
 
-    def first_seen(self, match_field) -> datetime.date:
-        return parser.parse(deep_get(self.noise, match_field, "first_seen"))
-
-    def last_seen(self, match_field) -> datetime.date:
-        return parser.parse(deep_get(self.noise, match_field, "last_seen_timestamp"))
-
-    def asn(self, match_field) -> str:
-        return deep_get(self.noise, match_field, "metadata", "asn")
-
-    def category(self, match_field) -> str:
-        return deep_get(self.noise, match_field, "metadata", "category")
-
-    def city(self, match_field) -> str:
-        return deep_get(self.noise, match_field, "metadata", "city")
-
-    def country(self, match_field) -> str:
-        return deep_get(self.noise, match_field, "metadata", "country")
-
-    def country_code(self, match_field) -> str:
-        return deep_get(self.noise, match_field, "metadata", "country_code")
-
-    def organization(self, match_field) -> str:
-        return deep_get(self.noise, match_field, "metadata", "organization")
-
-    def operating_system(self, match_field) -> str:
-        return deep_get(self.noise, match_field, "metadata", "os")
-
-    def region(self, match_field) -> str:
-        return deep_get(self.noise, match_field, "metadata", "region")
-
-    def is_tor(self, match_field) -> bool:
-        return deep_get(self.noise, match_field, "metadata", "tor")
-
-    def rev_dns(self, match_field) -> str:
-        return deep_get(self.noise, match_field, "metadata", "rdns")
-
-    def is_spoofable(self, match_field) -> str:
-        return deep_get(self.noise, match_field, "spoofable")
-
-    def tags_list(self, match_field) -> list:
-        tags = deep_get(self.noise, match_field, "tags")
+    def first_seen(self, match_field: str) -> datetime.date:
+        time = self._lookup(match_field, "first_seen")
+        if not time:
+            return None
+        if isinstance(time, list):
+            length = len(time)
+            if length == 0:
+                return None
+            if length == 1:
+                return parser.parse(time)
+            min_t = parser.parse(time[0])
+            for list_t in time[1:]:
+                list_t_parsed = parser.parse(list_t)
+                if list_t_parsed < min_t:
+                    min_t = list_t_parsed
+            return min_t
```

Comment:
nit: maybe something like 
```
        time = self._lookup(match_field, "first_seen")
        if not time:
            return None
        if isinstance(time, str):
            time = [time]

        return min(parser.parse(t) for t in time)
```

Reply:
Doesn't the `parser.parse(t) for t in time` alloc a list?

Reply:
Updated. Good suggestion!